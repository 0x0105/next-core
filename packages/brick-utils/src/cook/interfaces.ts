import { Expression } from "@babel/types";

export type PrecookScope = Set<string>;

export interface PrecookVisitorState {
  currentScope: PrecookScope;
  closures: PrecookScope[];
  attemptToVisitGlobals: Set<string>;
  identifierAsLiteralString?: boolean;
  collectParamNamesOnly?: string[];
}

export interface PrecookResult {
  source: string;
  expression: Expression;
  attemptToVisitGlobals: Set<string>;
}

export type CookScope = Map<string, CookScopeRef>;

export type CookScopeRef = {
  initialized: boolean;
  cooked?: any;
};

export interface CookVisitorState<T = any> {
  source: string;
  currentScope: CookScope;
  closures: CookScope[];
  identifierAsLiteralString?: boolean;
  spreadAsProperties?: boolean;
  collectParamNamesOnly?: string[];
  cookParamOnly?: boolean;
  argReceived?: any;
  optionalRef?: {
    ignored?: boolean;
  };
  memberCooked?: {
    object: ObjectCooked;
    property: PropertyCooked;
  };
  cooked?: T;
}

export type PropertyCooked = string | number;
export type PropertyEntryCooked = [PropertyCooked, any];
export type ObjectCooked = Record<PropertyCooked, any>;

export type VisitorCallback<T> = (node: any, state: T) => void;

export type VisitorFn<T> = (
  node: any,
  state: T,
  callback: VisitorCallback<T>
) => void;
