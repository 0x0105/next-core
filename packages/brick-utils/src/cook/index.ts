export { cook } from "./cook";
export { precook } from "./precook";
