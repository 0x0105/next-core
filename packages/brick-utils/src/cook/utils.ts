import {
  VisitorFn,
  PrecookVisitorState,
  CookVisitorState,
  PrecookScope,
  CookScope
} from "./interfaces";

export function walkFactory<T>(
  visitor: Record<string, VisitorFn<T>>,
  catchUnsupportedNodeType: (node: any) => void
): (node: any, state: T) => void {
  return function walk(node: any, state: T) {
    const nodeVisitor = visitor[node.type];
    if (nodeVisitor) {
      nodeVisitor(node, state, walk);
    } else {
      catchUnsupportedNodeType(node);
    }
  };
}

export function spawnPrecookState(
  parentState: PrecookVisitorState,
  extendsState?: Omit<
    PrecookVisitorState,
    "currentScope" | "closures" | "attemptToVisitGlobals"
  >
): PrecookVisitorState {
  return {
    currentScope: parentState.currentScope,
    closures: parentState.closures,
    attemptToVisitGlobals: parentState.attemptToVisitGlobals,
    ...extendsState
  };
}

export function spawnCookState(
  parentState: CookVisitorState,
  extendsState?: Omit<CookVisitorState, "source" | "currentScope" | "closures">
): CookVisitorState {
  return {
    source: parentState.source,
    currentScope: parentState.currentScope,
    closures: parentState.closures,
    ...extendsState
  };
}

export function getScopes(state: PrecookVisitorState): PrecookScope[];
export function getScopes(state: CookVisitorState): CookScope[];
export function getScopes(
  state: PrecookVisitorState | CookVisitorState
): (PrecookScope | CookScope)[] {
  return [state.currentScope].concat(state.closures);
}
