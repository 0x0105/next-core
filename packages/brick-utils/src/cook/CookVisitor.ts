import {
  ArrayExpression,
  ArrowFunctionExpression,
  BinaryExpression,
  CallExpression,
  ConditionalExpression,
  Identifier,
  LogicalExpression,
  MemberExpression,
  ObjectExpression,
  ObjectProperty,
  OptionalCallExpression,
  OptionalMemberExpression,
  SequenceExpression,
  SpreadElement,
  StringLiteral,
  TemplateLiteral,
  UnaryExpression
} from "@babel/types";
import {
  VisitorFn,
  CookVisitorState,
  CookScope,
  PropertyEntryCooked,
  ObjectCooked,
  PropertyCooked
} from "./interfaces";
import { spawnCookState, getScopes } from "./utils";

const CookVisitor: Record<string, VisitorFn<CookVisitorState>> = {
  ArrayExpression(
    node: ArrayExpression,
    state: CookVisitorState<any[]>,
    callback
  ) {
    const cookedElements = [];
    for (const element of node.elements) {
      if (element === null) {
        throw new SyntaxError(
          `Sparse arrays are not allowed: \`${state.source.substring(
            node.start,
            node.end
          )}\``
        );
      }
      const elementState = spawnCookState(state);
      callback(element, elementState);
      if (element.type === "SpreadElement") {
        cookedElements.push(...elementState.cooked);
      } else {
        cookedElements.push(elementState.cooked);
      }
    }
    state.cooked = cookedElements;
  },
  ArrowFunctionExpression(
    node: ArrowFunctionExpression,
    state: CookVisitorState<Function>,
    callback
  ) {
    if (!node.expression) {
      throw new SyntaxError(
        `Only an \`Expression\` is allowed in \`ArrowFunctionExpression\`'s body, but received: \`${state.source.substring(
          node.start,
          node.end
        )}\``
      );
    }

    if (node.async) {
      throw new SyntaxError(
        `Async function is not allowed, but received: \`${state.source.substring(
          node.start,
          node.end
        )}\``
      );
    }

    const cookedParamNames: string[] = [];
    for (const param of node.params) {
      const paramState: CookVisitorState<string> = spawnCookState(state, {
        identifierAsLiteralString: true
      });
      callback(
        param.type === "AssignmentPattern"
          ? param.left
          : param.type === "RestElement"
          ? param.argument
          : param,
        paramState
      );
      cookedParamNames.push(paramState.cooked);
    }

    state.cooked = function(...args: any[]) {
      const currentScope: CookScope = new Map();
      const bodyState: CookVisitorState = {
        currentScope,
        closures: getScopes(state),
        source: state.source
      };

      // For function parameters, define the current scope first.
      for (const paramName of cookedParamNames) {
        currentScope.set(paramName, {
          initialized: false
        });
      }

      node.params.forEach((param, index) => {
        const argReceived =
          param.type === "RestElement" ? args.slice(index) : args[index];
        const ref = currentScope.get(cookedParamNames[index]);
        if (param.type === "AssignmentPattern") {
          if (argReceived === undefined) {
            const paramValueState = spawnCookState(bodyState);
            callback(param.right, paramValueState);
            ref.cooked = paramValueState.cooked;
          } else {
            ref.cooked = argReceived;
          }
        } else {
          ref.cooked = argReceived;
        }
        ref.initialized = true;
      });

      callback(node.body, bodyState);
      return bodyState.cooked;
    };
  },
  BinaryExpression(node: BinaryExpression, state, callback) {
    const leftState = spawnCookState(state);
    callback(node.left, leftState);
    const leftCooked = leftState.cooked;

    const rightState = spawnCookState(state);
    callback(node.right, rightState);
    const rightCooked = rightState.cooked;

    switch (node.operator) {
      case "+":
        state.cooked = leftCooked + rightCooked;
        return;
      case "-":
        state.cooked = leftCooked - rightCooked;
        return;
      case "/":
        state.cooked = leftCooked / rightCooked;
        return;
      case "%":
        state.cooked = leftCooked % rightCooked;
        return;
      case "*":
        state.cooked = leftCooked * rightCooked;
        return;
      case "==":
        state.cooked = leftCooked == rightCooked;
        return;
      case "===":
        state.cooked = leftCooked === rightCooked;
        return;
      case "!=":
        state.cooked = leftCooked != rightCooked;
        return;
      case "!==":
        state.cooked = leftCooked !== rightCooked;
        return;
      case ">":
        state.cooked = leftCooked > rightCooked;
        return;
      case "<":
        state.cooked = leftCooked < rightCooked;
        return;
      case ">=":
        state.cooked = leftCooked >= rightCooked;
        return;
      case "<=":
        state.cooked = leftCooked <= rightCooked;
        return;
    }

    throw new SyntaxError(
      `Unsupported binary operator \`${
        node.operator
      }\`: \`${state.source.substring(node.start, node.end)}\``
    );
  },
  CallExpression(node: CallExpression, state, callback) {
    const calleeState: CookVisitorState<Function> = spawnCookState(state);
    callback(node.callee, calleeState);

    const cookedArgs = [];
    for (const arg of node.arguments) {
      const argState = spawnCookState(state);
      callback(arg, argState);
      if (arg.type === "SpreadElement") {
        cookedArgs.push(...argState.cooked);
      } else {
        cookedArgs.push(argState.cooked);
      }
    }

    if (typeof calleeState.cooked !== "function") {
      throw new TypeError(
        `${state.source.substring(
          node.callee.start,
          node.callee.end
        )} is not a function`
      );
    }

    const thisArg =
      node.callee.type === "MemberExpression" ||
      node.callee.type === "OptionalMemberExpression"
        ? calleeState.memberCooked.object
        : null;

    state.cooked = calleeState.cooked.apply(thisArg, cookedArgs);
  },
  ConditionalExpression(node: ConditionalExpression, state, callback) {
    const testState = spawnCookState(state);
    callback(node.test, testState);

    if (testState.cooked) {
      const consequentState = spawnCookState(state);
      callback(node.consequent, consequentState);
      state.cooked = consequentState.cooked;
    } else {
      const alternateState = spawnCookState(state);
      callback(node.alternate, alternateState);
      state.cooked = alternateState.cooked;
    }
  },
  Identifier(node: Identifier, state: CookVisitorState<string>) {
    if (state.identifierAsLiteralString) {
      state.cooked = node.name;
      return;
    }

    const scopes = getScopes(state);
    for (const scope of scopes) {
      if (scope.has(node.name)) {
        const ref = scope.get(node.name);
        if (!ref.initialized) {
          throw new ReferenceError(
            `Cannot access '${node.name}' before initialization`
          );
        }
        state.cooked = ref.cooked;
        return;
      }
    }

    throw new ReferenceError(`${node.name} is not defined`);
  },
  Literal(node: StringLiteral, state) {
    state.cooked = node.value;
  },
  LogicalExpression(node: LogicalExpression, state, callback) {
    const leftState = spawnCookState(state);
    callback(node.left, leftState);

    const leftCooked = leftState.cooked;

    switch (node.operator) {
      case "||":
        if (leftCooked) {
          state.cooked = leftCooked;
          return;
        }
        break;
      case "&&":
        if (!leftCooked) {
          state.cooked = leftCooked;
          return;
        }
        break;
      case "??":
        if (leftCooked !== null && leftCooked !== undefined) {
          state.cooked = leftCooked;
          return;
        }
        break;
    }

    const rightState = spawnCookState(state);
    callback(node.right, rightState);

    state.cooked = rightState.cooked;
  },
  MemberExpression(node: MemberExpression, state, callback) {
    const objectState: CookVisitorState<ObjectCooked> = spawnCookState(state);
    callback(node.object, objectState);

    const propertyState: CookVisitorState<PropertyCooked> = spawnCookState(
      state,
      {
        identifierAsLiteralString: !node.computed
      }
    );
    callback(node.property, propertyState);

    state.memberCooked = {
      object: objectState.cooked,
      property: propertyState.cooked
    };

    state.cooked = objectState.cooked[propertyState.cooked];
  },
  ObjectExpression(
    node: ObjectExpression,
    state: CookVisitorState<ObjectCooked>,
    callback
  ) {
    const cookedEntries: PropertyEntryCooked[] = [];
    for (const prop of node.properties) {
      const propState: CookVisitorState<
        PropertyEntryCooked | PropertyEntryCooked[]
      > = spawnCookState(state, {
        spreadAsProperties: true
      });
      callback(prop, propState);
      const propCooked = propState.cooked;
      if (prop.type === "SpreadElement") {
        if (propCooked !== null && propCooked !== undefined) {
          cookedEntries.push(
            ...Object.entries(propCooked as PropertyEntryCooked[])
          );
        }
      } else {
        cookedEntries.push(propCooked as PropertyEntryCooked);
      }
    }
    state.cooked = Object.fromEntries(cookedEntries);
  },
  OptionalCallExpression(node: OptionalCallExpression, state, callback) {
    const calleeState: CookVisitorState<Function> = spawnCookState(state, {
      optionalRef: {}
    });
    callback(node.callee, calleeState);

    const calleeCooked = calleeState.cooked;
    if (
      (node.optional || calleeState.optionalRef.ignored) &&
      (calleeCooked === null || calleeCooked === undefined)
    ) {
      state.cooked = undefined;
      if (state.optionalRef) {
        state.optionalRef.ignored = true;
      }
      return;
    }

    const cookedArgs = [];
    for (const arg of node.arguments) {
      const argState = spawnCookState(state);
      callback(arg, argState);
      if (arg.type === "SpreadElement") {
        cookedArgs.push(...argState.cooked);
      } else {
        cookedArgs.push(argState.cooked);
      }
    }

    if (typeof calleeCooked !== "function") {
      throw new TypeError(
        `${state.source.substring(
          node.callee.start,
          node.callee.end
        )} is not a function`
      );
    }

    const thisArg =
      node.callee.type === "MemberExpression" ||
      node.callee.type === "OptionalMemberExpression"
        ? calleeState.memberCooked.object
        : null;

    state.cooked = calleeCooked.apply(thisArg, cookedArgs);
  },
  OptionalMemberExpression(node: OptionalMemberExpression, state, callback) {
    const objectState: CookVisitorState<ObjectCooked> = spawnCookState(state, {
      optionalRef: {}
    });
    callback(node.object, objectState);

    const objectCooked = objectState.cooked;
    const objectCookedIsNil =
      objectCooked === null || objectCooked === undefined;
    if (
      (node.optional || objectState.optionalRef.ignored) &&
      objectCookedIsNil
    ) {
      state.cooked = undefined;
      if (state.optionalRef) {
        state.optionalRef.ignored = true;
      }
      return;
    }

    const propertyState: CookVisitorState<PropertyCooked> = spawnCookState(
      state,
      {
        identifierAsLiteralString: !node.computed
      }
    );
    callback(node.property, propertyState);

    const propertyCooked = propertyState.cooked;
    state.memberCooked = {
      object: objectCooked,
      property: propertyCooked
    };

    if (objectCookedIsNil) {
      throw new TypeError(
        `Cannot read property '${propertyCooked}' of ${objectCooked}`
      );
    }

    state.cooked = objectCooked[propertyCooked];
  },
  Property(
    node: ObjectProperty,
    state: CookVisitorState<PropertyEntryCooked>,
    callback
  ) {
    const keyState: CookVisitorState<PropertyCooked> = spawnCookState(state, {
      identifierAsLiteralString: !node.computed
    });
    callback(node.key, keyState);

    const valueState = spawnCookState(state);
    callback(node.value, valueState);

    state.cooked = [keyState.cooked, valueState.cooked];
  },
  SequenceExpression(node: SequenceExpression, state, callback) {
    let expressionState: CookVisitorState;
    for (const expression of node.expressions) {
      expressionState = spawnCookState(state);
      callback(expression, expressionState);
    }
    state.cooked = expressionState.cooked;
  },
  SpreadElement(node: SpreadElement, state, callback) {
    const argumentState = spawnCookState(state);
    callback(node.argument, argumentState);
    const cooked = argumentState.cooked;
    if (!state.spreadAsProperties && !isIterable(cooked)) {
      throw new TypeError(
        `${typeof cooked} is not iterable: \`${state.source.substring(
          node.start,
          node.end
        )}\``
      );
    }
    state.cooked = cooked;
  },
  TemplateLiteral(node: TemplateLiteral, state, callback) {
    let index = 0;
    const chunk: string[] = [node.quasis[index].value.cooked];
    for (const expression of node.expressions) {
      const expressionState = spawnCookState(state);
      callback(expression, expressionState);
      chunk.push(String(expressionState.cooked));
      chunk.push(node.quasis[(index += 1)].value.cooked);
    }
    state.cooked = chunk.join("");
  },
  UnaryExpression(node: UnaryExpression, state, callback) {
    const argumentState = spawnCookState(state);
    callback(node.argument, argumentState);

    const argumentCooked = argumentState.cooked;

    switch (node.operator) {
      case "!":
        state.cooked = !argumentCooked;
        return;
      case "+":
        state.cooked = +argumentCooked;
        return;
      case "-":
        state.cooked = -argumentCooked;
        return;
      case "typeof":
        state.cooked = typeof argumentCooked;
        return;
      case "void":
        state.cooked = undefined;
        return;
    }

    throw new SyntaxError(
      `Unsupported unary operator \`${
        node.operator
      }\`: \`${state.source.substring(node.start, node.end)}\``
    );
  }
};

function isIterable(cooked: any): boolean {
  if (cooked === null || cooked === undefined) {
    return false;
  }
  return typeof cooked[Symbol.iterator] === "function";
}

export default CookVisitor;
