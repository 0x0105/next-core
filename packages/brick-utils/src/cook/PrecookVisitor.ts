import {
  ArrayExpression,
  ArrowFunctionExpression,
  BinaryExpression,
  CallExpression,
  ConditionalExpression,
  Identifier,
  LogicalExpression,
  MemberExpression,
  ObjectExpression,
  ObjectProperty,
  OptionalCallExpression,
  OptionalMemberExpression,
  SequenceExpression,
  SpreadElement,
  TemplateLiteral,
  UnaryExpression
} from "@babel/types";
import { VisitorFn, PrecookVisitorState, PrecookScope } from "./interfaces";
import { spawnPrecookState, getScopes } from "./utils";

const PrecookVisitor: Record<string, VisitorFn<PrecookVisitorState>> = {
  ArrayExpression(node: ArrayExpression, state, callback) {
    for (const element of node.elements) {
      // ðŸš«Sparse arrays are not allowed.
      if (element !== null) {
        const elementState = spawnPrecookState(state);
        callback(element, elementState);
      }
    }
  },
  ArrowFunctionExpression(node: ArrowFunctionExpression, state, callback) {
    if (!node.expression) {
      // ðŸš«Only an `Expression` is allowed in `ArrowFunctionExpression`'s body.
      return;
    }

    const cookedParamNames: string[] = [];
    for (const param of node.params) {
      const paramState = spawnPrecookState(state, {
        identifierAsLiteralString: true
      });
      callback(
        param.type === "AssignmentPattern"
          ? param.left
          : param.type === "RestElement"
          ? param.argument
          : param,
        paramState
      );
      cookedParamNames.push(paramState.cookedIdentifierName);
    }

    const currentScope: PrecookScope = new Set(cookedParamNames);
    const bodyState: PrecookVisitorState = {
      currentScope,
      closures: getScopes(state),
      attemptToVisitGlobals: state.attemptToVisitGlobals
    };

    for (const param of node.params) {
      if (param.type === "AssignmentPattern") {
        const paramValueState = spawnPrecookState(bodyState);
        callback(param.right, paramValueState);
      }
    }

    callback(node.body, bodyState);
  },
  BinaryExpression(node: BinaryExpression, state, callback) {
    const leftState = spawnPrecookState(state);
    callback(node.left, leftState);

    const rightState = spawnPrecookState(state);
    callback(node.right, rightState);
  },
  CallExpression(node: CallExpression, state, callback) {
    const calleeState = spawnPrecookState(state);
    callback(node.callee, calleeState);

    for (const arg of node.arguments) {
      const argState = spawnPrecookState(state);
      callback(arg, argState);
    }
  },
  ConditionalExpression(node: ConditionalExpression, state, callback) {
    const testState = spawnPrecookState(state);
    callback(node.test, testState);

    const consequentState = spawnPrecookState(state);
    callback(node.consequent, consequentState);

    const alternateState = spawnPrecookState(state);
    callback(node.alternate, alternateState);
  },
  Identifier(node: Identifier, state) {
    if (state.identifierAsLiteralString) {
      state.cookedIdentifierName = node.name;
      return;
    }

    const scopes = getScopes(state);
    for (const scope of scopes) {
      if (scope.has(node.name)) {
        return;
      }
    }

    state.attemptToVisitGlobals.add(node.name);
  },
  Literal() {
    // Do nothing.
  },
  LogicalExpression(node: LogicalExpression, state, callback) {
    const leftState = spawnPrecookState(state);
    callback(node.left, leftState);

    const rightState = spawnPrecookState(state);
    callback(node.right, rightState);
  },
  MemberExpression(node: MemberExpression, state, callback) {
    const objectState = spawnPrecookState(state);
    callback(node.object, objectState);

    const propertyState = spawnPrecookState(state, {
      identifierAsLiteralString: !node.computed
    });
    callback(node.property, propertyState);
  },
  ObjectExpression(node: ObjectExpression, state, callback) {
    for (const prop of node.properties) {
      const propState: PrecookVisitorState = spawnPrecookState(state);
      callback(prop, propState);
    }
  },
  OptionalCallExpression(node: OptionalCallExpression, state, callback) {
    const calleeState = spawnPrecookState(state);
    callback(node.callee, calleeState);

    for (const arg of node.arguments) {
      const argState = spawnPrecookState(state);
      callback(arg, argState);
    }
  },
  OptionalMemberExpression(node: OptionalMemberExpression, state, callback) {
    const objectState = spawnPrecookState(state);
    callback(node.object, objectState);

    const propertyState = spawnPrecookState(state, {
      identifierAsLiteralString: !node.computed
    });
    callback(node.property, propertyState);
  },
  Property(node: ObjectProperty, state, callback) {
    const keyState = spawnPrecookState(state, {
      identifierAsLiteralString: !node.computed
    });
    callback(node.key, keyState);

    const valueState: PrecookVisitorState = spawnPrecookState(state);
    callback(node.value, valueState);
  },
  SequenceExpression(node: SequenceExpression, state, callback) {
    let expressionState: PrecookVisitorState;
    for (const expression of node.expressions) {
      expressionState = spawnPrecookState(state);
      callback(expression, expressionState);
    }
  },
  SpreadElement(node: SpreadElement, state, callback) {
    const argumentState = spawnPrecookState(state);
    callback(node.argument, argumentState);
  },
  TemplateLiteral(node: TemplateLiteral, state, callback) {
    for (const expression of node.expressions) {
      const expressionState = spawnPrecookState(state);
      callback(expression, expressionState);
    }
  },
  UnaryExpression(node: UnaryExpression, state, callback) {
    const argumentState = spawnPrecookState(state);
    callback(node.argument, argumentState);
  }
};

export default PrecookVisitor;
