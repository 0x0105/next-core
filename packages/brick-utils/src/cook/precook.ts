import { Node } from "@babel/types";
import { parseExpression } from "@babel/parser";
import { walkFactory } from "./utils";
import PrecookVisitor from "./PrecookVisitor";
import { PrecookVisitorState, PrecookResult } from "./interfaces";

export function precook(source: string): PrecookResult {
  const state: PrecookVisitorState = {
    currentScope: new Set(),
    closures: [],
    attemptToVisitGlobals: new Set()
  };
  const expression = parseExpression(source, {
    plugins: ["estree"]
  });
  walkFactory(PrecookVisitor, (node: Node) => {
    // eslint-disable-next-line no-console
    console.warn(
      `Unexpected node type \`${node.type}\`: \`${source.substring(
        node.start,
        node.end
      )}\``
    );
  })(expression, state);
  return {
    source,
    expression,
    attemptToVisitGlobals: state.attemptToVisitGlobals
  };
}
