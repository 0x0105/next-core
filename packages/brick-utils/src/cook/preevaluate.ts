import { PrecookOptions, precook } from "./precook";
import { PrecookResult } from "./interfaces";

// `raw` should always be asserted to `isEvaluable`.
export function preevaluate(
  raw: string,
  options?: PrecookOptions
): PrecookResult {
  const trimmed = raw.trim();
  const source = trimmed.substring(3, trimmed.length - 3);
  try {
    return precook(source, options);
  } catch (error) {
    throw new SyntaxError(`${error.message}, in "${raw}"`);
  }
}

export function isEvaluable(raw: string): boolean {
  return /^\s*<%\s/.test(raw) && /\s%>\s*$/.test(raw);
}
