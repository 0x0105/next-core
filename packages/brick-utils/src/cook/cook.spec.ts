import { cook } from "./cook";
import { precook } from "./precook";

describe("cook", () => {
  const getGlobalVariables = (): Record<string, any> => ({
    DATA: {
      for: "good",
      null: null,
      undefined: undefined,
      true: true,
      false: false,
      number5: 5,
      objectA: {
        onlyInA: 1,
        bothInAB: 2,
      },
      objectB: {
        onlyInB: 3,
        bothInAB: 4,
      },
    },
  });

  it.each<[string, any]>([
    ["'good'", "good"],
    ["1", 1],
    ["null", null],
    ["undefined", undefined],
    ["true", true],
    ["NaN", NaN],
    ["isNaN(NaN)", true],
    ["isNaN({})", true],
    ["isNaN(1)", false],
    ["Array.isArray([])", true],
    ["Array.isArray({})", false],
    ["Object.keys(DATA.objectA)", ["onlyInA", "bothInAB"]],
    ["String([1, 2])", "1,2"],
    ["Boolean(1)", true],
    ["Boolean(0)", false],
    ["Number(true)", 1],
    ["Number.isNaN({})", false],
    ["JSON.parse('[1]')", [1]],
    ["isFinite(5)", true],
    ["isFinite(Infinity)", false],
    ["parseFloat('5.3good')", 5.3],
    ["parseInt('5.3good')", 5],
    ["Math.max(1, 2)", 2],
    ["PIPES.string(null)", ""],
    ["location.href", "http://localhost/"],
    ["DATA.for", "good"],
    ["DATA['for']", "good"],
    ["DATA.other", undefined],
    ["{}", {}],
    [
      "{ quality: DATA.for, [DATA.true]: 'story' }",
      {
        quality: "good",
        true: "story",
      },
    ],
    ["[]", []],
    ["[1, DATA.number5]", [1, 5]],
    [
      // `ArrowFunctionExpression` mixed `CallExpression`
      "(a => a.b)({b: 'c'})",
      "c",
    ],
    [
      // `MemberExpression`
      "DATA.number5.toFixed(1)",
      "5.0",
    ],
    [
      // `OptionalMemberExpression`
      "DATA.number5?.toFixed(1)",
      "5.0",
    ],
    ["DATA.null || 'oops'", "oops"],
    ["DATA.for || 'oops'", "good"],
    ["DATA.null && 'oops'", null],
    ["DATA.for && 'oops'", "oops"],
    ["DATA.for ?? 'oops'", "good"],
    ["DATA.false ?? 'oops'", false],
    ["DATA.null ?? 'oops'", "oops"],
    ["DATA.undefined ?? 'oops'", "oops"],
    ["DATA.for?.length", 4],
    ["String?.(null)", "null"],
    ["DATA.notExisted?.length", undefined],
    ["DATA.notExisted?.length?.oops", undefined],
    ["DATA.notExisted?.()", undefined],
    ["DATA.notExisted?.()?.()", undefined],
    ["DATA.notExisted?.()?.length", undefined],
    ["DATA.notExisted?.length?.()", undefined],
    ["DATA.notExisted?.length.oops", undefined],
    ["DATA.notExisted?.length()", undefined],
    ["DATA.notExisted?.()()", undefined],
    ["!DATA.null", true],
    ["+DATA.true", 1],
    ["-DATA.true", -1],
    ["typeof DATA.for", "string"],
    ["void DATA.for", undefined],
    ["DATA.number5 + 1", 6],
    ["DATA.number5 - 1", 4],
    ["DATA.number5 / 2", 2.5],
    ["DATA.number5 % 2", 1],
    ["DATA.number5 * 2", 10],
    ["DATA.number5 == '5'", true],
    ["DATA.number5 == 4", false],
    ["DATA.number5 === 5", true],
    ["DATA.number5 === '5'", false],
    ["DATA.number5 != '5'", false],
    ["DATA.number5 != 4", true],
    ["DATA.number5 !== 5", false],
    ["DATA.number5 !== '5'", true],
    ["DATA.number5 > 4", true],
    ["DATA.number5 > 5", false],
    ["DATA.number5 < 6", true],
    ["DATA.number5 < 5", false],
    ["DATA.number5 >= 4", true],
    ["DATA.number5 >= 5", true],
    ["DATA.number5 >= 6", false],
    ["DATA.number5 <= 6", true],
    ["DATA.number5 <= 5", true],
    ["DATA.number5 <= 4", false],
    ["DATA.for ? 'yep': 'oops'", "yep"],
    ["DATA.null ? 'yep': 'oops'", "oops"],
    [
      // `SequenceExpression`
      "DATA.for, DATA.number5",
      5,
    ],
    [
      "`${null},${undefined},${true},${false},${{}},${[]},${[1,2]}${5},${NaN}`",
      `${null},${undefined},${true},${false},${{}},${[]},${[1, 2]}${5},${NaN}`,
    ],
    ["_.get(DATA, 'for')", "good"],
    ["[1,2,3].slice(1)", [2, 3]],
    ["[1, ...DATA.for, 2]", [1, "g", "o", "o", "d", 2]],
    ["[-1].concat(0, ...[1, 2], 3)", [-1, 0, 1, 2, 3]],
    ["[-1]?.concat(0, ...[1, 2], 3)", [-1, 0, 1, 2, 3]],
    [
      "{...DATA.objectA, ...DATA.objectB}",
      {
        onlyInA: 1,
        bothInAB: 4,
        onlyInB: 3,
      },
    ],
    ["{...null, ...undefined}", {}],
    ["[1, undefined].map((i = 5) => i)", [1, 5]],
    ["[1, undefined].map((i = DATA.number5) => i)", [1, 5]],
    [
      // `j` is not defined, but not evaluated either.
      "[1, 2].map((i = j + 1) => i)",
      [1, 2],
    ],
    [
      // `j` is not initialized, but not evaluated either.
      "[1, 2].map((i = j + 1, j) => i)",
      [1, 2],
    ],
    ["[1, 2].map(i => ((i, j = i + 1) => i + j)(i))", [3, 5]],
    [
      "[1, 2].map((...args) => args)",
      [
        [1, 0, [1, 2]],
        [2, 1, [1, 2]],
      ],
    ],
    [
      "[1, 2].map((i, ...args) => args)",
      [
        [0, [1, 2]],
        [1, [1, 2]],
      ],
    ],
  ])("cook(precook(%j), {...}) should return %j", (input, cooked) => {
    expect(cook(precook(input), getGlobalVariables())).toEqual(cooked);
  });

  it.each<string>([
    "this.bad",
    "() => {}",
    // `_.reverse` is not supplied.
    "_.reverse([0,1,2])",
    "Object.assign(DATA, { override: true })",
    // When `EVENT` is not supplied.
    "EVENT.detail",
    "delete DATA.for",
    "DATA.number5 ^ 1",
    "DATA.number5()",
    "DATA?.()",
    "DATA?.number5()",
    "DATA?.number5.notExisted.oops",
    "DATA.number5?.toFixed.oo.ps",
    "DATA.number5.toFixed?.().oops()",
    "[1, , 2]",
    "[...DATA]",
    "[...null]",
    "[...undefined]",
    // `j` is not defined, and will be evaluated.
    "[undefined].map((i = j + 1) => i)",
    // `j` is not initialized, and will be evaluated.
    "[undefined].map((i = j + 1, j) => i)",
    // Parameters affect the scopes.
    "[undefined].map((i = DATA.number5, DATA) => i)",
    "async () => null",
  ])("cook(precook(%j), {...}) should throw", (input) => {
    expect(() =>
      cook(precook(input), getGlobalVariables())
    ).toThrowErrorMatchingSnapshot();
  });
});
