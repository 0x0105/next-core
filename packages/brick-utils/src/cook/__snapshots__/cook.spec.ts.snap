// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cook cook(precook("() => {}"), {...}) should throw 1`] = `"Only an \`Expression\` is allowed in \`ArrowFunctionExpression\`'s body, but received: \`() => {}\`"`;

exports[`cook cook(precook("[...DATA]"), {...}) should throw 1`] = `"object is not iterable: \`...DATA\`"`;

exports[`cook cook(precook("[...null]"), {...}) should throw 1`] = `"object is not iterable: \`...null\`"`;

exports[`cook cook(precook("[...undefined]"), {...}) should throw 1`] = `"undefined is not iterable: \`...undefined\`"`;

exports[`cook cook(precook("[1, , 2]"), {...}) should throw 1`] = `"Sparse arrays are not allowed: \`[1, , 2]\`"`;

exports[`cook cook(precook("[undefined].map((i = DATA.number5, DATA) => i)"), {...}) should throw 1`] = `"Cannot access 'DATA' before initialization"`;

exports[`cook cook(precook("[undefined].map((i = j + 1) => i)"), {...}) should throw 1`] = `"j is not defined"`;

exports[`cook cook(precook("[undefined].map((i = j + 1, j) => i)"), {...}) should throw 1`] = `"Cannot access 'j' before initialization"`;

exports[`cook cook(precook("_.reverse([0,1,2])"), {...}) should throw 1`] = `"_.reverse is not a function"`;

exports[`cook cook(precook("DATA.number5 ^ 1"), {...}) should throw 1`] = `"Unsupported binary operator \`^\`: \`DATA.number5 ^ 1\`"`;

exports[`cook cook(precook("DATA.number5()"), {...}) should throw 1`] = `"DATA.number5 is not a function"`;

exports[`cook cook(precook("DATA.number5.toFixed?.().oops()"), {...}) should throw 1`] = `"DATA.number5.toFixed?.().oops is not a function"`;

exports[`cook cook(precook("DATA.number5?.toFixed.oo.ps"), {...}) should throw 1`] = `"Cannot read property 'ps' of undefined"`;

exports[`cook cook(precook("DATA?.()"), {...}) should throw 1`] = `"DATA is not a function"`;

exports[`cook cook(precook("DATA?.number5()"), {...}) should throw 1`] = `"DATA?.number5 is not a function"`;

exports[`cook cook(precook("DATA?.number5.notExisted.oops"), {...}) should throw 1`] = `"Cannot read property 'oops' of undefined"`;

exports[`cook cook(precook("EVENT.detail"), {...}) should throw 1`] = `"EVENT is not defined"`;

exports[`cook cook(precook("Object.assign(DATA, { override: true })"), {...}) should throw 1`] = `"Object.assign is not a function"`;

exports[`cook cook(precook("async () => null"), {...}) should throw 1`] = `"Async function is not allowed, but received: \`async () => null\`"`;

exports[`cook cook(precook("delete DATA.for"), {...}) should throw 1`] = `"Unsupported unary operator \`delete\`: \`delete DATA.for\`"`;

exports[`cook cook(precook("this.bad"), {...}) should throw 1`] = `"Unexpected node type \`ThisExpression\`: \`this\`"`;
